tests/pytests/unit/states/test_hg.py

```diff
@@ -0,0 +1,120 @@
+"""
+    :codeauthor: Rahul Handay <rahulha@saltstack.com>
+"""
+
+import os
+
+import pytest
+import salt.states.hg as hg
+from tests.support.mock import MagicMock, patch
+
+
+@pytest.fixture
+def configure_loader_modules():
+    return {hg: {}}
+
+
+def test_latest():
+    """
+    Test to Make sure the repository is cloned to the given directory and is up to date
+    """
+    ret = {"changes": {}, "comment": "", "name": "salt", "result": True}
+    mock = MagicMock(return_value=True)
+    with patch.object(hg, "_fail", mock):
+        assert hg.latest("salt")
+
+    mock = MagicMock(side_effect=[False, True, False, False, False, False])
+    with patch.object(os.path, "isdir", mock):
+        mock = MagicMock(return_value=True)
+        with patch.object(hg, "_handle_existing", mock):
+            assert hg.latest("salt", target="c:\\salt")
```

Comment:
**suggestion**: prefer using `autospec=True` in patch, and not manually creating `MagicMock`. Pretty sure this could be written as `with patch.object(os.path, "isdir", autospec=True, side_effect=[False, True, False, False, False, False]):`, etc.

Reply:
This suggestion is broadly applicable, but it would especially simplifiy this test.